Improved README instructions for building in Termux on Android devices

JackJollimore opened the pull request:
Added direction to build with `make`, simplified `OpenBlas`, and `CLBlast` installation method. Clarified where to _move_ models in Termux. `llama.cpp` is diverse, so I added a `main` Usage example for new users.

These instructions should work for any Android device capable of running `llama.cpp`.

Commit:
Improve README.md for building in Termux on Android devices

Added direction to build with `make`, and simplified `OpenBlas` method. These instructions should work for any Android device capable of running `llama.cpp`.

Clarified on where to _move_ the models in Termux. `llama.cpp` is diverse, so I added a Usage example for new users.

I left `CLBlast` instructions unchanged for now.

Commit:
clarified wording

JackJollimore commented:
I'm open to suggeation. Also, tell me if there's any problems with these changes.

Commit:
clean up Usage example

made Usage example more clear in the case Users build with cmake

Commit:
dummy-proof directions

$HOME before make is unneccesary.

If User is new with Termux then typing cd storage/downloads before $HOME is confusing.

clarified directions.

Cebtenzzre commented:
What is $HOME meant to do? My Termux has the bash shell so running it as a command would just print an error.

JackJollimore commented:
> What is $HOME meant to do? My Termux has the bash shell so running it as a command would just print an error.

Thanks for pointing that out. I expected $HOME to navigate to `/data/data/com.termux/files/home` in Termux.

I can just add the direct path, but let me look and see if there's a bash method.

Commit:
corrected navigation

$HOME is a script within fish shell, assuming User uses bash then we use `cd` to navigate to /data/data/com.termux/files/home

JackJollimore commented:
That was an oversight on my part :) 

Using `cd` in place of $HOME fixes it.

staviq commented:
If I remember correctly, android "sandboxes" apps such that native home folder is the app folder, and I checked with termux, not sure about other apps, but standard unix `~` works in termux, and resolves to `/data/data/com.termux/files/home`

Commit:
Clarify move model directions

remove reference to Termux `$HOME` path.

JackJollimore commented:
> I would just say "termux app directory ( which is automatically mapped to home directory )" or something like that.

I think Termux app directory is less clear as it might refer to `~/storage/shared/android/data/com.Termux`, but I removed `$HOME`  because my intention is to make sure the User moves the model to the correct spot.

Thank you.

Cebtenzzre commented:
> I removed `$HOME` because my intention is to make sure the User moves the model to the correct spot.

I think it would be sufficient to say "to the home directory (`~/`)". The example doesn't use its full path, and the user shouldn't have to know or care what the absolute path technically is.

Commit:
crystal-clear move path destination

User needs to move the model to the correct location, and doesn't need to know/care about the absolute path in Termux

Commit:
typo fixed

JackJollimore commented:
Looks clean and streamlined to me:

![Screenshot_20230828-103150_Iceraven](https://github.com/ggerganov/llama.cpp/assets/130917767/13b5ce3b-c13e-49f7-984d-995a881cdf28)

JackJollimore commented:
For any that are interested, here's a possible installation method for `CLBlast`:
```
pkg install ocl-icd opencl-headers opencl-clhpp clinfo
cd
git clone https://github.com/CNugteren/CLBlast
cd CLBlast
cmake -B build \
  -DBUILD_SHARED_LIBS=OFF \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_INSTALL_PREFIX=/data/data/com.termux/files/usr
cd build
make -j8
make install
```

Build with `CLBlast`:
```
cd
cd llama.cpp
cmake -B build -DLLAMA_CLBLAST=ON
cd build
cmake --build . --config Release
```

Run on GPU:
```
GGML_OPENCL_PLATFORM=0
GGML_OPENCL_DEVICE=0
export LD_LIBRARY_PATH=/vendor/lib64:$LD_LIBRARY_PATH
```

then `./llama.cpp/build/bin/main -m ~/7b-model.gguf --color -c 2048 --keep -1 -n -2 -b 10 -i -ins -ngl 1`. Some Android devices require `LD_LIBRARY_PATH=/system/vendor/lib64` before `main` `


Admittedly, trying to ensure `OpenCL` enables for other peoples devices is tricky, so fool-proofing is tough. It's a WIP.

Commit:
Update Usage Example

original example was generic, adjusted --prompt to fit alpaca-style

Commit:
Update README.md

typo fixed

Commit:
Made Usage example more easy to copy

also looks better, and consistent

JackJollimore commented:
@Cebtenzzre Are there any other inconsistencies, or changes neccesary here?
![Screenshot_20230831-110712_Iceraven](https://github.com/ggerganov/llama.cpp/assets/130917767/858cfb29-6ec7-4165-a1d2-6fd55fb571ea)

Cebtenzzre commented:
> Are there any other inconsistencies, or changes neccesary here?

Have you seen my review comments?

JackJollimore commented:
> Have you seen my review comments?

Hmm, I don't see a review comment, other than a couple days ago: https://github.com/ggerganov/llama.cpp/pull/2840#issuecomment-1694937991 - I implemenred the change here: https://github.com/ggerganov/llama.cpp/pull/2840/commits/0a7ef3a5f9e164b3a70239bdd8b4d6f9061f2962

Is there something else I'm missing?

Cebtenzzre commented:
> Hmm, I don't see a review comment

Do you see them [here](https://github.com/ggerganov/llama.cpp/pull/2840/files)?

JackJollimore commented:
> Do you see them [here](https://github.com/ggerganov/llama.cpp/pull/2840/files)?

No. I looked over the [entire page](https://github.com/ggerganov/llama.cpp/pull/2840/files), even desktop-mode and CTRL F.

![Screenshot_20230831_202656](https://github.com/ggerganov/llama.cpp/assets/130917767/4de097c0-6f5c-4d13-886f-dac311c9d460)

Github says you're requested for review, but it doesn't mention a comment, otherwise I'd check it out:
![Screenshot_20230831_202943](https://github.com/ggerganov/llama.cpp/assets/130917767/598e29af-6cdd-4433-9040-5a3a3eff9bd7)

Commit:
adjust OpenBlas from cmake to make

JackJollimore commented on README.md:
````diff
@@ -791,17 +791,41 @@ Finally, copy the `llama` binary and the model files to your device storage. Her
+
+Alternatively, to enable CLBlast then install the requisite OpenCL packages:
 ```
 apt install ocl-icd opencl-headers opencl-clhpp clinfo
````
I'm open to change `CLBlast` instructions - these were introduced here: https://github.com/ggerganov/llama.cpp/pull/1828#pullrequestreview-1484741361, but they don't work on my device, so I don't use them.

I found this is ineffecient when `llama.cpp` has  a new release:
```
cp /data/data/com.termux/files/usr/include/openblas/cblas.h .
cp /data/data/com.termux/files/usr/include/openblas/openblas_config.h .
```

For my device, `LD_LIBRARY_PATH=/vendor/lib64:$PREFIX/lib` is required before `main` due to Termux/[Adreno GPU](https://github.com/termux/termux-packages/issues/16852#issuecomment-1579047943)

But for many devices, Adding `export LD_LIBRARY_PATH=/vendor/lib64` to ~/.bashrc is certainly better.

I figure `consider documenting this final part in a .sh script file.` is obscure, so I'd be glad to change it.

Commit:
Clean/Update Android ClBlast section

Replaced outdated, ineffecient steps for a streamlined method.

Commit:
typo fix

JackJollimore commented on README.md:
````diff
@@ -791,48 +791,51 @@ Finally, copy the `llama` binary and the model files to your device storage. Her
+```
 
-For easy and swift re-execution, consider documenting this final part in a .sh script file. This will enable you to rerun the process with minimal hassle.
+Finally, use `export LD_LIBRARY_PATH=/vendor/lib64:$LD_LIBRARY_PATH` to enable GPU, then `./main ... -ngl 1`
````
@Cebtenzzre Everything else looks good to me. Line 836 is the only thing that I dislike esthetically, but I'm picky.

I'm open to changing the wording, or format, so let me know what you think.

Commit:
easily copy/paste export

made export better for copy/pasting, looks better

Commit:
stylized OpenBlas text

now is correctly matched with `CLBlast` & `llama.cpp`

JackJollimore commented:
The goal of this PR is to be universal and compatible, so here's a Note from current `master`:  (some Android devices, like the Zenfone 8, need the following command instead - "export LD_LIBRARY_PATH=/system/vendor/lib64:$LD_LIBRARY_PATH".), then it links to reddit.

Some devices path from `LD_LIBRARY_PATH=/vendor/`, whereas others path from `LD_LIBRARY_PATH=/system/` to /lib64

`export` from the wrong path == fail, so I think it makes sense to note the alternative for compatibility purposes.

Commit:
Ensure device compatibility for CLBlast

Some devices path from Vendor, others path from System

Commit:
fix typo
